base/abstractarray.jl

```diff
@@ -97,51 +123,75 @@ linearindexing(::LinearIndexing, ::LinearIndexing) = LinearSlow()
 end
 
 # check along a single dimension
-checkbounds(::Type{Bool}, sz::Integer, i) = throw(ArgumentError("unable to check bounds for indices of type $(typeof(i))"))
-checkbounds(::Type{Bool}, sz::Integer, i::Real) = 1 <= i <= sz
-checkbounds(::Type{Bool}, sz::Integer, ::Colon) = true
-function checkbounds(::Type{Bool}, sz::Integer, r::Range)
+checkindex(::Type{Bool}, inds::UnitRange, i) = throw(ArgumentError("unable to check bounds for indices of type $(typeof(i))"))
+checkindex(::Type{Bool}, inds::UnitRange, i::Real) = (first(inds) <= i) & (i <= last(inds))
```

Comment:
Hmm, `i in r::Range` should check that `i` is an element of `r`. This is checking that `i` lies in the interval spanned by the range, which is different if `i` is, e.g., 3.2 and `r` is `1:5`. So now I think we shouldn't call `in`.

But I'll take the chance to special-case `in` with this implementation anyway; no need to perform division or invoke `mod` for a `UnitRange`.